src/tensor/tensor.cpp

```diff
@@ -0,0 +1,84 @@
+#include "./tensor/tensor.h"
+
+#include <algorithm>
+#include <cstddef>
+#include <iostream>
+#include <numeric>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+Shape::Shape(std::vector<size_t> dims) : dimensions(std::move(dims)) {
+  total_elements = std::accumulate(dimensions.begin(), dimensions.end(),
+                                   static_cast<size_t>(1),
+                                   [](size_t a, size_t b) { return a * b; });
+}
+
+size_t Shape::get_rank() const { return dimensions.size(); }
+
+template <typename T>
```

Comment:
Please, move template implementation to the header, otherwise it will not work for other data types specified by the user